pkg/types/chain_reader.go

```diff
@@ -41,8 +41,40 @@ type ChainReader interface {
 	GetLatestValue(ctx context.Context, contractName string, method string, params, returnVal any) error
 
 	// Bind will override current bindings for the same contract, if one has been set and will return an error if the
-	// contract is not known by the ChainReader, or if the address is invalid
+	// contract is not known by the ChainReader, or if the Address is invalid
 	Bind(ctx context.Context, bindings []BoundContract) error
+	// TODO UnBind?, doesn't seem like its needed?
+
+	//TODO ReBindByKey, UnBindByKey, this is needed in some places where addresses change but the contract stays the same
+	// ReBindByKey(ctx context.Context, key, address string)
+	// UnBindByKey() or UnBindByKey(ctx context.Context, key, address string)
+
+	QueryKey(ctx context.Context, keys string, queryFilter QueryFilter, limitAndSort LimitAndSort, sequenceDataType any) ([]Sequence, error)
+	QueryKeys(ctx context.Context, keys []string, queryFilter QueryFilter, limitAndSort LimitAndSort, sequenceDataTypes []any) ([][]Sequence, error)
+	QueryByKeyValuesIn(ctx context.Context, keyValueIn KeyValueIn, queryFilter QueryFilter, limitAndSort LimitAndSort, sequenceDataType any) ([]Sequence, error)
+	QueryByKeysValuesIn(ctx context.Context, keysValuesIn []KeyValueIn, queryFilter QueryFilter, limitAndSort LimitAndSort, sequenceDataTypes []any) ([][]Sequence, error)
+	QueryByKeyValueEquality(ctx context.Context, KeyValuesByEquality KeyValuesByEquality, queryFilter QueryFilter, limitAndSort LimitAndSort, sequenceDataType any) ([]Sequence, error)
+	QueryByKeysValuesEquality(ctx context.Context, KeysValuesByEquality []KeyValuesByEquality, queryFilter QueryFilter, limitAndSort LimitAndSort, sequenceDataTypes []any) ([][]Sequence, error)
+
+	// TODO make EVM words map to a key and then do this through the QueryByEquality methods.
+	// GetCommitReportMatchingSeqNum()
+	// GetSendRequestsBetweenSeqNums()
+	// GetCommitReportGreaterThanSeqNum()
+}
+
+type ValueEquality struct {
+	Value    string
+	Operator ComparisonOperator
+}
```

Comment:
If `ComparisonOperator` can be anything (equality, inequality, negation, etc.) then why is it called `ValueEquality` and `QueryKeyValueEquality` rather than `ValueComparison` and `QueryKeyValueComparison`?

Reply:
Actually, I think the main suggestion I'd have for this whole interface is that it could be simplified a lot by just adding a primitive like this:

```
type KeyValuePrimitive struct {
	Key string
	Value string
	Operator ComparisonOperator
}
```
Where the effect of it is to say "KeyValueMap[Key] = Value" (for equality) or eg "KeyValueMap[Key] > Value" for Gt, etc.

If we add that, then I think there is no need to have 6 different versions of `QueryKey` there just needs to be one, or maybe two. We wouldn't need singular and plural for each though.

For example, if you want to use 2 different keys to filter the results, instead of using `QueryByKeysEquality` you'd just pass a filter expression to `QueryByKeys` (or whatever we decide to call it) that's a Bool AND type, with each of its children a `KeyValuePrimitive` enforcing that these two keys equal particular values.  Similarly, instead of using `QueryByKeyValuesIn` you could either AND together a `KeyValuePrimitive` for each value the key can match, or better yet just define another primitive like this:

```
type KeyValueInPrimitive struct {
	Key string
	Values []string
}
```
which would mean KeyValueMap[Key] = one of the values in the set of Values

It's late, and off the top of my head I'm not sure quite what to do with the other 2 methods (`QueryKey(s)`) but it seems like they could probably also be dealt with in the same way.

Reply:
1. Fair point on naming
2. It can't be a primitive and be part of the Query Filter interface. I am keeping keys separate from QueryFilter on purpose for a couple of reasons: 
- Its impossible to query for multiple keys with the same QueryFIlter. From what I've seen the usual pattern for querying for multiple keys (multiple eventSigs) is to query for them in the same way (same time range, block range etc ...). If you put keys in the QueryFilter you cannot turn that into sensible db query because everything is jumbled up together. This is similar to how ORMs work, you provide what you are filtering for separately eg. `Where(a.Equal(x), b.Gte(y)...).Fetch(SomeData)`. So providing the key separately is necessary and is less complex.
- Another point is that it provides a better UX because it makes it is obvious that the key must be provided whereas if it was part of QueryFilter it can't be enforced at compile time.
3. Having separate Querying methods makes it way less complex on the implementation. Each of the methods serves a different purpose. QueryKey/s is just requesting sequences by keys and the other 4 are for requesting sequences by keys and providing a way to filter over data stored in the sequence. I can see merging `QueryByKeyValuesIn` and `QueryByKeysValuesEquality`, but couldn't think of a clean way to express that with proto